extensions/shared/upgrade-nudge/index.jsx

```diff
@@ -1,49 +1,64 @@
 /**
  * External dependencies
  */
-import { get, startsWith } from 'lodash';
 import { __, sprintf } from '@wordpress/i18n';
 import { addQueryArgs } from '@wordpress/url';
 import { Button } from '@wordpress/components';
+import { Component } from '@wordpress/element';
+import { get, startsWith } from 'lodash';
 import { withSelect } from '@wordpress/data';
 import Gridicon from 'gridicons';
 
 /**
  * Internal dependencies
  */
+import analytics from 'lib/analytics';
```

Comment:
See https://github.com/Automattic/jetpack/pull/12965/files#r300639854

Reply:
I'm hesitant to pull this lib into the `/extensions` build 😕 

I know there's unguarded `window` usage which I'd like to see improved. Adding `_inc/client` to the extensions build opens a door to pull in anything and I don't think it's a good precedent to allowing anything from the older client app.

I'd rather audit this a bit and pull it out into a clean shared location and set the precedent that client modules can't be pulled into extensions without further thought.

Reply:
Understandable. It might be available in window global tho — that's how [wpcom/Jetpack Gutenberg bridge](https://github.com/Automattic/wp-calypso/blob/18e0409c61c6bd0e391cac398d10b2c77bf21df2/apps/wpcom-block-editor/src/common/track-search.js) is accessing it.

We talked with @mtias  to have a look at Tracking in block editor anyway, so I'm pretty sure we'll loop to this quite soon even if we'd implement something that's not ideal as a first step. p1562316269414600-slack-serenity

Reply:
@sirreal @ockham do you have thoughts what would be a good location in Jetpack repo to move `_inc/client/lib/analytics` to so that it would be easy to share between block editor stuff in `/extensions` and dashboard stuff in `_inc/client`?

Reply:
Just for the record, whatever location we decide to put those files needs to be synced to WP.com by Fusion so that the build also works there. (I don't think that would currently the case with anything in `_inc/client/`).

Reply:
The _North Star_ (ideal soluttion informing interim steps) should probably be an analytics npm published from the Calypso repo, since (AFAIK) Jetpack's analytics module was also forked from Calypso at the time.

Reply:
I agree with both sentiments. I think on this round we can get at least location for easy syncing nailed.

I'll raise NPM question in paYJgx-5g-p2

Reply:
The work of extracting the analytics module was started here, but it doesn't seem to have made a lot of progress. I'm not sure whether that was due to a significant issue or otherwise:

https://github.com/Automattic/wp-calypso/pull/33234

@scruffian were there major problems extracting the analytics library?

---

> do you have thoughts what would be a good location in Jetpack repo to move `_inc/client/lib/analytics` to so that it would be easy to share between block editor stuff in `/extensions` and dashboard stuff in `_inc/client`?

Interesting question 🤔 @ockham has a good point that we should think about using a shared library in the future. We should also consider whether we'll be doing more and more of this in Jetpack and whether a solid precedent needs to be established.

Right now we're building several block editor extensions out of the `/extensions` directory and a single app out of `_inc/client` with no (good, established) way to share code. In Calypso we're in the process of going from a single monolithic app to a bunch of apps (under `/apps`) that depend on some related, isolated, and shared modules (under `/packages`) and a bunch of shared code mixed up with the main Calypso application (under `/client`). Ideally, the Calypso app would also move under `/apps`, and shared code could continue to live in a shared location, but I suspect it would benefit from becoming more principled like the other packages.

- We can't use `/packages`, that's where imported PHP packages reside.
- We can't use `/modules`, that's where Jetpack modules reside. This would be rather nice but 🤷‍♂️
-  Maybe `/shared-modules` or similar is best?

It's also worth considering what expectations should be set for the shared code. The spectrum ranges from code soup to true node modules. I'd propose that, at the very least, it should be structured _like_ modules, where the top level consists of directories that contain an `index.(js|jsx|ts|tsx|json)` that is the only way shared code should be accessed.

Reply:
> @scruffian were there major problems extracting the analytics library?

Ben is AFK, let's ask @Automattic/team-calypso since I think they followed the process?

Reply:
> > @scruffian were there major problems extracting the analytics library?
> 
> Ben is AFK, let's ask @Automattic/team-calypso since I think they followed the process?

Not sure it's high up on their radar.

FWIW, Ben's PR's desc seems to have a rather comprehensive list of dependencies, a lot of which are non-trivial to sort out.

Ironically, AFAIK Ben was doing this as preparatory work to be able to move Calypso's `UpgradeNudge` (which uses analytics) to `packages/` in order to publish it as an npm for re-use in Jetpack -- for the Paid Block Nudge. I kinda followed that path and concluded it was too much work for the time being.

We _might_ want to invest a little more time (one hour?) to assess whether Calypso's `lib/analytics` code could be reorganized a it to lend itself more easily to re-use.

For the time being, and in order to not get blocked, we might need an interim solution in Jetpack. This should be ideally well understood by all stakeholders -- people need to know what to put there, and to what end.

I agree with Jon's deliberations on nomenclature for that path. I'll just point out that `shared-modules/` is prone to be understood as something related to (PHP) `modules/`.

Maybe just a root `shared/` folder? Whatever it is, it needs a big fat README saying something like: this is something that's both used by `_inc/client/` (which is the JP React dashboard), and `extensions/` (which are Gutenberg blocks for both JP and WP.com), and that like the latter, it is synced to WP.com via Fusion.

Reply:
We'll likely need to strike a balance here.

I do believe that pulling things from `_inc` into `extensions` is a bad precedent and makes fusion messy.

If we consider this work "urgent," I'm willing to accept that debt and land if we can scope some time to work on a better fix like making an analytics package.

The alternative is likely moving this to a shared location, which is relatively quick, but will set create precedents we haven't thought through carefully and will remove any incentives to work on a shareable analytics package.

Reply:
>If we consider this work "urgent," 

Yep, the goal is to have an event tracked before we can publish the feature.

Started little something: https://github.com/Automattic/jetpack/pull/13031

Reply:
Hmm, does the component need to fire the track event directly? 

We control the full component here, so might be cleaner to create a custom analytics store + event to dispatch, and add related middleware handling that's served off of wp.widgets.com handled in a script somewhere in the common folder https://github.com/Automattic/wp-calypso/tree/master/apps/wpcom-block-editor/src/common

Reply:
>We might want to invest a little more time (one hour?) to assess whether Calypso's lib/analytics code could be reorganized a it to lend itself more easily to re-use.

Since Edit teams are looking into tracking, let's not step too much into refactoring just yet before we know what's the direction of work those teams will take. (paYJgx-5g-p2).

Reply:
>We control the full component here, so might be cleaner to create a custom analytics store + event to dispatch, and add related middleware handling that's served off of wp.widgets.com handled in a script somewhere in the common folder https://github.com/Automattic/wp-calypso/tree/master/apps/wpcom-block-editor/src/common

Good idea. Worth checking it out! Thanks.